### ISSUE TITLE
Add dynamite sudden death

### ISSUE THREAD

**Issue by rmarquis** [etlegacy/etlegacy#1498]
On maps where the final objective is to be dynamited (such as oasis anti tank guns), dynamites have to be planted before the timer reaches 30 seconds. Sometimes, dynamites are successfully planted, but everyone knows the game is lost because it's not been done early enough.

A nice solution is to implement some "sudden death" mechanics: if the dyna is planted after the 30 sec limits, then the game will not end until the dynamite either explodes or is disarmed - allowing both team to fight until the very end.

This is a feature that has been implemented in etpub, but also in Brink later on and works quite well.

> (g_dyno 1): If there is 30 seconds or less on the clock and dynamite is planted on an objective, the clock will continue to run past zero. The match will not end until the dynamite either explodes, or is disarmed. You are not allowed to plant additional dynamite during sudden death.

**Comment by rmarquis** [comment]
I'd suggest to implement the above by default, without any cvar.

Special care should be taken to ensure the implementation of the feature still make sense in the context of Stopwatch mode. If not possible, only implement it for other game modes.

**Comment by Aciz** [comment]
Comment from eMMiel since he doesn't have a github account regarding SW:


"So, when you play a braundorf match that takes 10 minutes. You were stressed for 10 minutes straight, trying to not let them plant. Then just at 9:50 they succeed in planting. So the match would go until 10:20.  The other guys would also need a time until 10:20. So that last match can go on until 10:50.  So you get an extra minute of match time. But a fullhold is 10 minutes and not 10:50 so, it can be that a team has to fullhold it 20 seconds longer. 20 sec is one allies spawn. So it would be completely unfair for that second time, cause they have one more spawn to plant that dynamite."

**Comment by mittermichal** [comment]
@emmiel this is about removing the unnecesary part when its decided that defending team won. 
What you are talking about is different idea - giving the team that did objective "too late" overtime.

**Comment by Crytiqal** [comment]
> Comment from eMMiel since he doesn't have a github account regarding SW:
> 
> "So, when you play a braundorf match that takes 10 minutes. You were stressed for 10 minutes straight, trying to not let them plant. Then just at 9:50 they succeed in planting. So the match would go until 10:20. The other guys would also need a time until 10:20. So that last match can go on until 10:50. So you get an extra minute of match time. But a fullhold is 10 minutes and not 10:50 so, it can be that a team has to fullhold it 20 seconds longer. 20 sec is one allies spawn. So it would be completely unfair for that second time, cause they have one more spawn to plant that dynamite."

If the 2nd team plants the bomb AFTER 9:50, say 9:55, their time would go to 10:25, making them take 5 seconds more for the map, thus losing..
The stopwatch time would have to be set on the time the last dynamite was planted, so on the 2nd turn, you MUST plant the bomb ON or before that time, be able to go into overtime but still beat the clock.
Roundtime should be (over)time minus 30 seconds.

Example:
1st team plant bomb on 09:50 --> win on 10:20
2nd team has 09:50 to plant the bomb, or they will lose. So the roundtime would be set on 09:50 (10:20-0:30) for them to plant the bomb.

Example 2:
1st team plant bomb on 09:50 --> defuse on 10:10 and lose...
2nd team has 10:10 - 30s = 09:40 to plant the bomb and win at 10:10 or get defused sooner and lose..
If the bomb would get defused at 09:45, even if they planted a 2nd bomb at 09:50, it would take longer to explode than 10:10 and be slower, thus still losing.

**Comment by Exagone313** [comment]
I'm working on this, I'll send a PR soon.

**Comment by rmarquis** [comment]
@Exagone313 Have you checked Stopwatch mode? Haven't tried on my side, but as it is, it probably breaks the clock. See @Crytiqal's comment above to have an idea of what should be done.

Also, the reason I'm not a fan of adding yet a new cvar is because of standardizing gameplay of Legacy mod across servers as much as possible, and that there is no downside of having it implemented by default (as long as stopwatch clock is done correctly). Over time, we implemented many existing mods options as default, as providing cvars just for the sake of it doesn't make much sense.

**Comment by Aranud** [comment]
My bad, I closed the ticket when the merge was done. Reopen.

**Comment by Exagone313** [comment]
I'll see about it when I get the time.

**Comment by Exagone313** [comment]
I didn't understand the comment about stopwatch but I have never played this game type, so I'll need to give it a try somehow locally.

About the cvar, as a player and potential server administrator, I wouldn't be happy to have a change of gameplay that can't be disabled. Another solution would be to enable this cvar by default and decide on a later version to remove it based on user input.

**Comment by rmarquis** [comment]
Yes, admins are never happy about any change but if we listened to them we'd had dozens or hundreds of cvars even when they don't make any sense. It looks like some of you think of this change as an apocalyptic event, but I fail to see how it isn't a 100% improvement over the current state.

Anyway, here's a quick recap, in a theoretical map that involves dynamiting the final objective in 10min:

Currently:

(Scenario 1)
* Round 1: Team A plants the dyna before 9:30, at let's say 9:20, Team X fails to disarm, the game is won at 9:50. The clock is set at 9:50.
* Round 2: Team X needs to plant dyna before 9:20, or they will loose even if the clock has still time (30 sec are useless)

(Scenario 2)
* Round 1: Team A fails to plant dyna, or plant dyna too late after 9:30, or team B disarm. Clock is set at 10:00. The last 30 sec are useless if Team A fails to plant before 9:30.
* Round 2: Team X needs to plant dyna before 9:30, or they will lose the round. The last 30 sec are also useless.

With Sudden Death:

(Scenario 1)
* Round 1: Team A plants the dyna before 9:30, at let's say 9:20. Team X fails to disarm, the game is won at 9:50. The clock is set at 9:20.
* Round 2: Team X needs to plant dyna before 9:20, or they will loose.

(Scenario 2)
* Round 1: Team A plants the dyna after 9:30, at let's say 9:50. The game will enter sudden death at 10:00, and ends at 10:20. Team X fails to disarm, the game is won at 10:20. The clock is set at 9:50.
* Round 2: Team X needs to plant dyna before 9:50. ~~Sudden death starts at 10:00.~~ Can't plant new dyna past 10:00.

(Scenario 3)
* Round 1: Team A plants the dyna after 9:30, at let's say 9:50. The game will enter sudden death at 10:00, and ends at 10:20. Team X succeed in disarming 20 sec later, Team A lose at 10:10. The clock is set at 9:40.
* Round 2: Team X needs to plant dyna before 9:40. ~~Sudden death starts at 10:00.~~ Can't plant new dyna past 10:00.

(Scenario 4)
* Round 1: Team A fails to plant dyna, or team B disarm. Clock is set at  9:30.
* Round 2: Team X needs to plant dyna before 9:30, or they will lose the round.

I *think* that covers all scenarios, hoping I haven't missed anything here.

So the clock is basically set 30 sec prior to what it in the current state, and new dyna can't be planted once time reaches sudden death extension. However, do note this *only* applies when the map requires the final objective to be destroyed. In all other case (retrieving objective, escorting truck, ...) the clock behaviour is unchanged.

**Comment by rmarquis** [comment]
Also, for those that might still skeptical: think of the above as being implemented as described, but instead of a timelimit of 10:00 (+potentially 1 to 30 secs of sudden death), imagine the timelimit being set at 9:30, with 1-30 sec of sudden death time added: This would be 100% vanilla bahaviour, only with the useless time being removed.

**Comment by Aciz** [comment]
The issue with those sudden death scenarios you listed is that it's then possible to win with a worse time, sudden death should NEVER occur on 2nd round of stopwatch when time is already set. On first round it makes sense, but not on second.

**Comment by rmarquis** [comment]
Yes, sorry. I rather meant "new dyna can't be planted" as to equilibrate on both round, but sudden death itself shouldn't happen on round 2.

**Comment by Helldembez** [comment]
About this mechanic in stopwatch:

In my opinion if you are using the suddendeath mechanic you are accepting that you have to fight until the last second. Doesn't matter if the 2nd round someone plants later than in the first round.

10 minutes is 10 minutes. When planted between 09:30 and 10:00 and the defending team does not manage to defuse it the attacker completed the objective. The only question is what time has to be beat. I would say the time to beat is still the maximum time of 10 minutes. The scoring would just be the opposite of a double fullhold.

Normally a fullhold gives makes it 1-0. Double fullhold 1-1.
Now with suddendeath mechanic it becomes 0-1 if the dynamite explodes in aftertime. If the other team does the same its 1-1 again. No need to count the time after the timelimit.

**Comment by Exagone313** [comment]
I'm seeing three possibilities by reading your responses:
* either disable sudden death on second round (`g_currentRound > 0`?)
* or, do not count the extra time "added" by sudden death (this way, any extra time is counted the same), but need extra care to tie-based sudden death
* or, disable sudden death on stopwatch.

Having to do some calculation based on if the dynamite is the final objective or not probably can't be done.

**Comment by rmarquis** [comment]
The primordial aspect is that the feature doesn't give any technical advantage to any side. I'd remove any "useless" time and adjust the clock accordingly, but as you can see the implementation is a bit convoluted but need to be rock solid. I guess a lot of testing will be required to "get it right", and that we might figure out unforeseen issues along the way.

I don't really know about the final objective issue, I hadn't thought about it. Maybe it's possible to evaluate with the map script.

This said, since this feature is already merged but breaks StopWatch as it is, I'd start by temporarily disabling it for this mode.

**Comment by Exagone313** [comment]
For the moment, the cvar is off by default.

**Comment by Exagone313** [comment]
I've been thinking of the algorithm to implement for sudden death on stopwatch.

There are two rounds in stopwatch:
* on the first round (`g_currentRound == 0`), time limit is set to the value given to `wm_set_round_timelimit` in map script
* on the second round (`g_currentRound == 1`), time limit is set to the value of `g_nextTimeLimit`

When the first round ends (also calculated on second round, but useless as far as I understand):
* if the winning team is the defender team, `g_nextTimeLimit` is set to the current time limit (`g_timelimit`)
* else, `g_nextTimeLimit` is set to `(timeCurrent - startTime) / 60000` (milliseconds are converted to a floating value in minutes)

The calculation of `g_nextTimeLimit` is done only for the first round, so this is probably the only thing than needs to be adjusted, but it may also be possible to add new variables between the two rounds.

When the dynamite is planted before the time limit without reaching sudden death, the behavior is to use the current time limit and it must not be altered.

When the round ends later due to **tie-based** sudden death, with the attacking team winning, the next time limit is set to a subtract of times, calculated like above. This behavior won't be changed either. But my understanding is that very few maps use this behavior, so it's not a big deal (for now?) if it doesn't play right with dynamite-based sudden death.

When dynamite-based sudden death is reached and that the attacking team wins (when the defender team wins, the behavior is unchanged), I see two possibilities:
* to set `g_nextTimeLimit` to `g_timelimit` on the first round, just like when the defender team wins:
    * on the second round, the *added* time will not be relevant, both teams will get the *same* chance to win (in my opinion), even when using dynamite-based sudden death
    * if tie-based sudden death is followed by dynamite-based sudden death, the behavior is not natural, this may need to be addressed later
* to disable dynamite-based sudden death on the second round
    * if it was used to win on the first round, `g_nextTimeLimit` will be greater than the default `g_timelimit`, dynamite will need to be planted 30 seconds before the end
    * else, the behavior is unchanged

I prefer the first solution, feel free to give your thoughts.

**Comment by rmarquis** [comment]
> The calculation of g_nextTimeLimit is done only for the first round, so this is probably the only thing than needs to be adjusted, but it may also be possible to add new variables between the two rounds.

Do note this has actually been changed recently in ed5a2bd8b63eb7079ecd6b4d63810bc26fa4b1db.

`g_nextTimeLimit` is now computed on both round, and is reset again at the end of intermission of round 2 (when SW match ends).

Also a quick note on `g_currentRound` that you might have missed: the value flips at the end of play, before intermission. So during intermission of the first round, it's value is `1` (and not zero), and conversely during intermission of the second round it is set to `0` (and not 1). I was quite surprised and I am not sure why they coded it this way nor try to "fix it" by fear of breaking more stuff, but that might be relevant here maybe. 

> But my understanding is that very few maps use this behavior

I am not sure what you mean here, care to expand? What do you mean by "tie-based sudden death" here?

I'll also try to have a deeper look after a good night sleep. This feature is actually quite complex and kinda makes my brain melt, and without sketching a few things on paper it's quite hard to follow.

**Comment by Exagone313** [comment]
I have seen the calculation was made on round 2 too, but I didn't understand why, thanks. I'm still not sure when each block of code are executed, I made the lazy choice not to get it thoroughly.

I haven't taken the time to check what it leads to, but tie-based sudden death refers to this:

https://github.com/etlegacy/etlegacy/blob/ab1f4cbd44f108f9d6b361c4a20a86846aeb6f65/src/game/g_main.c#L4277-L4282

https://github.com/etlegacy/etlegacy/blob/ab1f4cbd44f108f9d6b361c4a20a86846aeb6f65/src/game/g_main.c#L4138-L4151

I don't know when `CS_MULTI_MAPWINNER` is used.

**Comment by rmarquis** [comment]
Ah yes. It's very seldom used, only in a few maps where both team try to do objectives at the same time (baserace, Fata Morgana, Hydro Dam come to mind).

`CS_MULTI_MAPWINNER` is usually set by map scripts in the more conventional attacker/defender maps. The above maps use some sort of counters instead.

**Comment by Exagone313** [comment]
Should I try writing one or two draft pull requests? This could be used for making some test builds to check the behavior in-game.

**Comment by rmarquis** [comment]
Sure, go ahead.

**Comment by Exagone313** [comment]
I opened both pull requests, but I have not yet tested those in-game. I don't know if the CI can produce "trusted" mod builds, it would make it easier for that process.

**Comment by rmarquis** [comment]
CI only ensures the code compiles. The resulting code can still be jonky or even produce crashes.
Thanks for the drafts!

**Comment by x0rnn** [comment]
Shouldn't this be map/objective-specific also? What if you play fueldump, <30 seconds remain, Allies plant dynamite at the Axis MG tower or Axis CP and not at the fuel dump - sudden death activates anyway?

And maps with 2 dynamitable objectives (oasis, sw_oasis_b3, erdenberg_t2, tc_base), if neither of the objectives are already destroyed, then if one gets planted - then the sudden death is meaningless because the other can't be and allies can't win anyway.

**Comment by Aciz** [comment]
Yes, the current implementation is flawed and does not take into account on what the actual objective is, a dynamite planted on any objective will trigger a sudden death, no matter if the objective is the map-ending objective or not. It will even trigger in maps where the winning objective is not a dynamiteable objective, e.g. sudden death would trigger on Radar if allies have a dynamite planted on some Axis MG nest.

**Comment by x0rnn** [comment]
I made a lua for "sudden death" mode: https://github.com/x0rnn/etlegacy-lua-scripts/blob/master/suddendeath.lua (modified a dynamite timer lua script), Quake 3's modified sound file (should play after "Dynamite planted" sound): https://github.com/x0rnn/etpro/blob/master/lua/sudden_death.wav

Can be made to work on SW by checking g_currentRound I guess and not allowing sudden death in 2nd round. Edit: works on SW now too.

**Comment by x0rnn** [comment]
> Also a quick note on g_currentRound that you might have missed: the value flips at the end of play, before intermission. So during intermission of the first round, it's value is 1 (and not zero), and conversely during intermission of the second round it is set to 0 (and not 1). I was quite surprised and I am not sure why they coded it this way nor try to "fix it" by fear of breaking more stuff, but that might be relevant here maybe.

Does this mean it's 0 during the first round, flips to 1 during intermission, stays 1 during the second round and then flips back to 0 at intermission of second round?

Edit: yes.

**Comment by Exagone313** [comment]
Hello, I finally took some time to test my patches on #1834 and #1835, which are about the stopwatch game mode.

One thing troubled me when trying sudden death again: it should only activate when the attacking team planted a dynamite. Having sudden death when the defending team planted a dynamite (e.g. at an enemy command post) is useless. This should be fixed first.

Checking if the sudden death dynamite was planted on the final objective of the map could be done in a future patch. I think it is not an easy work to be able to support all the maps out of the box.

Back to the patches:
* With #1835, if the attacking team uses sudden death in the first round to win, the second round will have increased time (at most 30 seconds, the dynamite time) but sudden death will always be disabled in second rounds.
    * It feels really odd to me that the second round would get more than the time limit in the first place, I think that sudden death shouldn't add any time.
    * When sudden death happens, the time countdown is (graphically) stopped at the end, so if this is part of the game play, it should be possible to use it in any round.
* With #1834, if the attacking team uses sudden death to win, the saved time will be the time limit, there will be no added time.
    * It feels natural, sudden death is part of the game play and it happens "after" the countdown has stopped, so the time "added" doesn't have any meaning.
    * There is just the message `Objective NOT reached in time` displayed in the console after the second round when both teams made the exact same score, which happens either when the defending team wins or when the attacking team wins by sudden death. But it's not a big deal, it can stay as-is (it's just a console message anyway).

I would recommend to:
* make a patch in `DynamiteOnObjective` function so that sudden death only works if a dynamite was planted by the attacking team
* add the patch made in #1834 
* keep thinking about how to make the dynamite check more relevant to the final objective of the map

What do you think?

**Comment by Exagone313** [comment]
The code that disables waiting in map script may not be working as expected, because the sudden death flag is removed after the dynamite explodes. Pretty sure this makes the attacking team to loose in maps where it waits and plays an animation after the dynamite exploded before ending the match. This should be revised by not removing the sudden death flag when the attacking team successfully destroyed the objective with their dynamite.

**Comment by x0rnn** [comment]
There are also maps that have 2 final dynamiteable objectives (oasis, sw_oasis_b3, tc_base, erdenberg_t2). Sudden death shouldn't activate if at least one of those hasn't already been destroyed.

**Comment by x0rnn** [comment]
> About this mechanic in stopwatch:
> 
> In my opinion if you are using the suddendeath mechanic you are accepting that you have to fight until the last second. Doesn't matter if the 2nd round someone plants later than in the first round.
> 
> 10 minutes is 10 minutes. When planted between 09:30 and 10:00 and the defending team does not manage to defuse it the attacker completed the objective. The only question is what time has to be beat. I would say the time to beat is still the maximum time of 10 minutes. The scoring would just be the opposite of a double fullhold.
> 
> Normally a fullhold gives makes it 1-0. Double fullhold 1-1. Now with suddendeath mechanic it becomes 0-1 if the dynamite explodes in aftertime. If the other team does the same its 1-1 again. No need to count the time after the timelimit.

This should be voted upon by the majority, because in my opinion if one team won in sudden death in faster time than the other, that team should win, because it was faster (*stopwatch*).

In my lua implementation I made it so:

Scenario 1:
Round 1:
Sudden death enabled by default:
Allies don't plant at all, match ends at 10:00
Round 2:
Sudden death still enabled by default:
a) Allies don't plant at all, match ends at 10:00, draw
b) Allies plant at 9:50, map goes in overtime, explodes at 10:20, Allies win
Both teams had sudden death available, first one failed, second succeeded - fair

Scenario 2:
Round 1:
Sudden death enabled by default:
Allies plant at 5:00, explodes & match ends at 5:30, Allies win
Round 2:
Sudden death disabled:
a) Allies don't plant at all, Axis win
b) Allies have time to plant to 4:59 or 5:00 at latest, map ends at 5:30
Sudden death was not needed/activated - fair
- Or do you enable sudden death and simply decrease the timelimit in 2nd round by 30s to avoid uselessly dragging the time after 5:00 if dyna wasn't planted? 

Scenario 3:
Round 1:
Sudden death enabled by default:
Allies plant at 9:50, explodes at 10:20 in overtime, Allies win
Round 2:
Sudden death disabled, just timelimit is set +30s more now:
a) Allies don't plant at all, Axis win
b) Allies have time to plant to 9:49 or 9:50 at latest, map ends at 10:20
Second team technically or semantically doesn't have "sudden death" but still has extra time to plant. Wouldn't be fair if second team "wins" map with sudden death at 10:30; they took longer. 
- Or do you enable sudden death and simply decrease the timelimit in 2nd round by 30s to avoid uselessly dragging the time after 9:50 if dyna wasn't planted? 

Scenario 4:
Round 1:
Sudden death enabled by default:
Allies plant at 9:50, defused at 10:10 in overtime, Axis win
Round 2:
Sudden death disabled, but timelimit is still set +30s more now:
a) Allies don't plant at all, Axis win
b) Allies have time to plant to 9:49 or 9:50 at latest, map ends at 10:20
Second team technically or semantically doesn't have "sudden death" but still has extra time to plant - fair.
- Or do you enable sudden death and set timelimit to default 10min, and Allies have time up to 9:59 to plant?

**Comment by Exagone313** [comment]
> There are also maps that have 2 final dynamiteable objectives (oasis, sw_oasis_b3, tc_base, erdenberg_t2). Sudden death shouldn't activate if at least one of those hasn't already been destroyed.

You could plant on both objectives at the same time, so that both would explode after the time limit, then sudden death should happen too.

**Comment by x0rnn** [comment]
I guess that is also a rare possibility, in this case I could make a threshold of maybe waiting <0.5 second if the 2nd dynamite is also planted, then enable sudden death. But if one is planted at 9:50, and other at 9:59 or even 9:51, it shouldn't activate.

Edit: or maybe it should? Another thing that should be voted on...

**Comment by rmarquis** [comment]
> Checking if the sudden death dynamite was planted on the final objective of the map could be done in a future patch. I think it is not an easy work to be able to support all the maps out of the box.

My memories are a bit fuzzy, but there is a way to differentiate "primary" and "secondary" objectives (see code related to #1003). Adding a check to ensure sudden death is enabled for primary objectives might do the trick here.

**Comment by oheikk** [comment]
> My memories are a bit fuzzy, but there is a way to differentiate "primary" and "secondary" objectives (see code related to https://github.com/etlegacy/etlegacy/issues/1003). Adding a check to ensure sudden death is enabled for primary objectives might do the trick here.

Primary objectives can be things you need to do in the map to advance, they don't necessarily need to be the final objective. Unfortunately there's no way to tell if a primary objective actually ends the map apart from parsing the script and the entities (and parsing these is akin to parsing some program code and trying to automatically figure out what it does.)

It's not even trivial to tell if the primary objective is something you need to dynamite, though you can follow the `target` value of the `trigget_objective_info` entity related to the Primary objective, see if it's a `func_explosive`, and if so look at its `scriptname` and then parse the script to see if it's set as `constructible_class 3`, though with the caveat that if some mapper decided to implement some logic that changes the `constructible_class` based on some condition, this can potentially break too.

Apart from all that, there's also the issue that not all mappers actually follow the primary/secondary objective conventions - all they do is give info in the limbo menu. You can make a perfectly functional map without doing all that, and most players won't even notice.

With all this in mind I'd recommend just providing some basic info in a file about maps that end with a dynamite (e.g. info about the entities that need to be blown up for the map to end, `scriptname` and/or `targetname` should do) in a format that server admins can easily edit if they want to support more maps. @x0rnn hard-coded them in his https://github.com/x0rnn/etlegacy-lua-scripts/blob/master/suddendeath.lua but I don't think it's necessary to go that far.

**Comment by x0rnn** [comment]
> With all this in mind I'd recommend just providing some basic info in a file about maps that end with a dynamite (e.g. info about the entities that need to be blown up for the map to end, `scriptname` and/or `targetname` should do) in a format that server admins can easily edit if they want to support more maps. @x0rnn hard-coded them in his https://github.com/x0rnn/etlegacy-lua-scripts/blob/master/suddendeath.lua but I don't think it's necessary to go that far.

That's the same as hard-coding them, though. You can modify the lua and it's the same as if you modified that file... Although a bit easier, I guess.

**Comment by Exagone313** [comment]
> The code that disables waiting in map script may not be working as expected

I made some tests for this but the code that disables `wait` is working as expected, because the round will be ending by `wm_endround` in scripts and will never reach `timelimit_hit`. The dynamite is still exploding when evaluating the `death` event for the dynamitable objective.

**Comment by x0rnn** [comment]
Honestly, there's so many different variables, checks and conditions that need to be considered to properly implement this, and not to mention all the potential outcomes that should be agreed upon beforehand.

Introducing sudden death into stopwatch mode complicates things quite a bit already, both technically and non-technically (the Scenarios I mentioned above), and when you get to maps with 2 dynamiteable objectives even more so, regardless of gametype. 

For example, let's assume a timelimit of 10, Oasis:
Neither Gun is destroyed already, dynamite is planted 9:50, you obviously don't enable sudden death. But what if both dynamites are planted at the same time (milliseconds apart)? What if one is planted 9:40 and the other 9:50? What if one is planted 9:20 and then the other at 9:45? A lazy "fix" is to just enable sudden death if and only if one objective is already destroyed, which is what I used in my lua so far, but I'll try to make it work for these cases too, and probably come across some other complications in the process...

Also, further dynamites need to be disabled during sudden death, but what if it's defused 9:45, then they need to be re-enabled and sudden death possible again. On that note, do you disable them at 10:00 or as soon as 1 dynamite is planted? Having ex. 3 dynamites to confuse Axis AND sudden death doesn't seem fair - they're already been "punished" with sudden death. Or do you allow just 1 sudden death without subsequent ones? Edit: I guess this is a bigger issue for lua because you have to literally extend the timelimit 30s, you can't just "freeze" it at 10:00 and then enable "sudden death"...

And another thing: 2 engineers could throw 2 dynamites on the floor, activate 1, sudden death gets enabled, other one gets activated 10s later - now what? The other one should be ignored or removed if possible. A simple fix would be also to take the pliers away.

**Comment by x0rnn** [comment]
Uips suggested extending sudden death to objective delivery maps: if obj is stolen before timelimit hit, enable SD until obj is either secured or returned.

**Comment by Mateos81** [comment]
Also have in mind double-obj maps such as Den of Lions where both teams attack/defend (aka no winner set)

**Comment by u1ps** [comment]
No idea how far this project is but would be awesome if it is fully working on normal game mode default settings (Oasis one team attacking last gun up triggering sudden death). 

Most of the discussion seems to be around SW more complicated settings.

on off feature for server to choose like friendly fire

**Comment by Exagone313** [comment]
> Also have in mind double-obj maps such as Den of Lions where both teams attack/defend (aka no winner set)

For the record, in my recently published patch, I checked if a team was defending, rather than if a team was not attacking, so that maps where both teams can win by attacking are exempted from rules that apply to defenders.

EDIT: Now that I think about it, it may not work properly in maps such as [V2 Factory](https://et.trackbase.net/map/39), where the Allies team is defending but can also win by attacking the factory in the early part of the map. Will need to check about it.

**Comment by x0rnn** [comment]
> Uips suggested extending sudden death to objective delivery maps: if obj is stolen before timelimit hit, enable SD until obj is either secured or returned.

Players might steal an objective and hide for an extended period of time... Happens on radar. Obj carrier hiding for minutes while rest of the team frags... Maybe should make a timelimit how long a player can carry it before he is gibbed.

**Comment by x0rnn** [comment]
https://github.com/x0rnn/etlegacy-lua-scripts/blob/master/suddendeath.lua

I finally got the time and willingness to finish this, so now it's compatible for both stopwatch mode and campaign/mapvote, as well as with maps with double dynamite objectives, pauses/unpauses and custom timebonuses. Compatible maps are: "battery, sw_battery, fueldump, braundorf_b4, etl_braundorf, mp_sub_rc1, sub2, sw_oasis_b3, oasis, tc_base, etl_base, erdenberg_t2"

For SW mode, the 2nd round will automatically deduct 30s if the dynamite exploded in the previous round, to prevent needlessly dragging the 2nd round for longer than needed. For example, if a map is 10 minutes long, if Allies plant at 5:00 and it explodes at 5:30, the 2nd round will be set to 5 minutes, the same if planted at 9:50 and it explodes in overtime at 10:20, the timer in the 2nd round will be set to 9:50. If, however, Axis defuse the dynamite at 10:19 in overtime, the map immediately ends and the timer in the 2nd round will be set to the default 10 minutes, allowing the other team to also extend the map up to 10:30 if they plant the dynamite at the last second.

For maps with double dynamite objectives, sudden death will get enabled if either 1 objective is already destroyed, or if 1 got planted with less than a minute remaining, and the other got planted with less than 30s remaining. If one gets defused, sudden death deactivates.

If dynamite gets defused while sudden death is activated in the regular map timelimit, planting a new dynamite will activate a new sudden death.

It is not possible to arm additional dynamites (at the objective) when sudden death activates.
